### ISSUE TITLE
TempImage throws std::bad_alloc under casacore-3.4 when handling degenerate dimensions

### ISSUE THREAD

**Issue by VoronkovMA** [casacore/casacore#1171]
I found this issue converting ASKAPsoft into casacore-3.4. Apparently, if an array with additional degenerate dimensions (w.r.t. the dimensions of the TempImage object) is passed to the TempImage::put method, the program  fails by either throwing a std::bad_alloc exception or because it got killed by the operating system. It is likely that it tries to allocate all available memory in a loop before the failure occurs. Interestingly, the same code works under casacore-3.2.1. 

Although passing an array with not exactly the same, but conforming, shape is a sloppy code (and the fix is trivial via Array::reform), it may be worth adding assert statements or otherwise check if the relevant code can be made more reliable, especially because there is a change in behaviour between casacore-3.4 and casacore-3.2.1. And moreover, the resulting bug is difficult to trace.

I attached some test code which can be used to reproduce the issue. It requires just casacore.   

[test.tar.gz](https://github.com/casacore/casacore/files/8158686/test.tar.gz)

**Comment by tammojan** [comment]
Thanks for the clear issue report and reproducer. This change of behavior could have been introduced with the arrays modernization that @aroffringa did some time ago. Perhaps he can have a look?

```
#include <casacore/images/Images/TempImage.h>
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/Matrix.h>
#include <casacore/coordinates/Coordinates/CoordinateSystem.h>
#include <casacore/coordinates/Coordinates/LinearCoordinate.h>


#include <iostream>

casacore::CoordinateSystem getDummyCoordinateSystem()
{
   // this is just to create a dummy coordinate system required in the constructor
   casacore::Vector<casacore::String> names(2);
   names[0]="x"; names[1]="y";
   casacore::Vector<double> increment(2 ,1.);

   casacore::Matrix<double> xform(2,2,0.);
   xform.diagonal() = 1.;
   const casacore::LinearCoordinate linear(names, casacore::Vector<casacore::String>(2,"pixel"),
           casacore::Vector<double>(2,0.),increment, xform, casacore::Vector<double>(2,0.));

   casacore::CoordinateSystem coords;
   coords.addCoordinate(linear);
   return coords;
}


int main() {
   try {
      const casacore::IPosition shape(2,1024,1024);

      // this works fine
      //const casacore::IPosition arrayShape(shape);

      // this results in std::bad_alloc or the program is killed
      const casacore::IPosition arrayShape(casacore::IPosition(4,1024,1024,1,1));

      casacore::Array<double> buf(arrayShape, 0.);

      const casacore::CoordinateSystem coords(getDummyCoordinateSystem());
      // +100 to ensure it is in memory, takes care of overheads
      const double maxMemoryInMB = double(shape.product()*sizeof(double))/1024./1024.+100;
      casacore::TempImage<double> tmpImg(casacore::TiledShape(shape), coords, maxMemoryInMB);
      tmpImg.put(buf);
      std::cout<<"test completed normally"<<std::endl;
   } catch (const std::exception &ex) {
      std::cerr<<ex.what()<<std::endl;
      return -1;
   }
   return 0;
}
```

**Comment by aroffringa** [comment]
The code leads in the end to a call of
```
template<class T>
void ArrayLattice<T>::doPutSlice (const Array<T>& sourceBuffer,
				  const IPosition& where, 
				  const IPosition& stride)
```
with `ldim=2`, `sdim=4`. It then calls
```
Array<T> allAxes(sourceBuffer.addDegenerate(ldim-sdim));
```
hence it tries to call `addDegenerate` with a negative number. The parameter of `addDegenerate` is a`size_t`; before the arrays rewrite it was a `uInt`. On first look, it looks like this code was therefore always incorrect, but the change from `uInt` to `size_t` causes it to allocate 2^64 objects instead of 2^32 objects.

I don't particularly use this part of casacore, nor do I understand it, so I'm not the best person to fix this.

**Comment by VoronkovMA** [comment]
Thanks. @aroffringa. May be it worked with casacore-3.2.1 because we may have had a spare 4 Gb of RAM. 

Perhaps, one could just use sourceBuffer.reform instead of addDegenerate here. I think it checks that the array conforms with the requested shape too. Or, if there are any objections to this approach, just an assert (that ldim >= sdim) could be added.

I also wonder if doGetSlice has the same issue.

**Comment by tammojan** [comment]
I think this was closed by #1196.
